src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,181 @@
+         * data.  Data is buffered even before you call this method, so be sure to
+         * explicitly wait for any and all data sent by the other side.
+         */
+        waitForDataWithAtLeastLength(length) {
+            if (pendingResolve) {
+                throw new Error("only one wait allowed at a time.");
+            }
+            if (receivedData && receivedData.length >= length) {
+                let promise = Promise.resolve(receivedData);
+                receivedData = null;
+                return promise;
+            }
+            console.log("(" + socketType + " waiting for " + length + " bytes)\n");
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+                wantDataLength = length;
+            });
+        },
+        waitForAnyDataAndClose() {
+            if (pendingResolve) {
+                throw new Error("only one wait allowed at a time.");
+            }
+
+            return new Promise(function(resolve, reject) {
+                pendingResolve = resolve;
+                // we may receive no data before getting close, in which case we want to
+                // return an empty array
+                receivedData = new Uint8Array();
+                wantDataAndClose = true;
+            });
+        },
+    };
+}
+
+var tcpsocket = class tcpsocket extends ExtensionAPI {
+    getAPI(context) {
+        return {
+            experiments: {
+                tcpsocket: {
+                    /** 
+                     * Send a DNS query stored in buf over a TCP socket to a 
+                     * nameserver addressed by addr
+                     */
+                    async sendDNSQuery(addr, buf) {
+                        let tcp_socket;
+                        let tcp_event_queue;
+                        let answer = new Uint8Array();
+                        try {
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
+                            tcp_event_queue = listenForEventsOnSocket(tcp_socket, "client");
+
+                            // Wait until the socket has opened a connection to the DNS server
+                            let nextEvent = (await tcp_event_queue.waitForEvent()).type;
+                            if (nextEvent == "open" && tcp_socket.readyState == "open") {
+                                console.log("client opened socket and readyState is open");
+                            } else {
+                                return {"error_code": 1, "data": answer};
+                            }
+                        } catch(e) {
+                            return {"error_code": 1, "data": answer};
+                        }
+
+                        try {
+                            tcp_socket.send(buf.buffer, buf.byteOffset, buf.byteLength);
+                            answer = await tcp_event_queue.waitForDataWithAtLeastLength(buf.byteLength);
```

Comment:
I don't have too much experience with `Promise` constructors, and it's not immediately clear to me how to simplify this to my needs. I'd like to talk this through with you on Slack

Reply:
> I don't have too much experience with `Promise` constructors, and it's not immediately clear to me how to simplify this API to my needs.

You've already used it in your `sleep` implementation.

In an `async function`, `await` can be used to "pause" the control flow of that function until the promise has been resolved. To use `await`, you need a promise. Callback and event-based APIs (such as the socket instance) don't return promises, but you can easily obtain a promise with the `Promise` constructor, like you already did in your sleep implementation.

> So first, it seems like what we want to do is open the socket, wait for the next event, and ensure it's an "open" event. If not, we throw an error. To do this, do we need to add an event listener for each possible event, and then somehow wait on a global `Promise` that each event listener can resolve? This is a similar pattern to the code I borrowed from the aforementioned test file, but I wonder if there's a simpler design.
>
> Second, it seems like we need to send a query, wait until we get some initial data, convert the first two entries from the Uint8array in the first "data" event into a 16-bit "length" integer (in which each entry in the array represents a byte), and keep waiting for data events until we get "length" bytes, concatenating Uint8arrays as necessary. Again, this design is similar to what the test code uses, but I have two questions:
> 
>     1. If we receive less than the expected number of bytes from the first "data" event and never get additional "data" events, then we'd be waiting forever, right? Presumably the TCPSocket interface would fire an "error" event if TCP gave up after waiting too long, but I'm not sure.

From your comments, I think that you're trying to wait for one event at a time. You don't need to limit yourself to that, it is very well possible to wait for multiple events (and/or callbacks). It would look like this:

```
let socket = somehowConstructASocket();
// TODO: Wrap in try-finally, and clean up the socket in the finally call.
// TODO: Use context.callOnClose to ensure that the socket is closed on extension unload,
//       + context.forgotOnClose to release the cleanup handler when the task finished
//         before the extension has unloaded.
await new Promise((resolve, reject) => {
  // This (arrow) function is immediately called by the Promise constructor.
  // The generated promise will settle as soon as resolve or reject is called.
  // After calling resolve or reject, subsequent calls to resolve/reject are ignored.

  let buffers = [];
  socket.ondata = (arrayBuffer) => {
    buffers.push(arrayBuffer);
    // TODO: check if buffers is complete and call resolve();
  };
  socket.onerror = (error) => {
    reject(new ExtensionError("meaningful error message here"));
  };

  // If the socket API doesn't offer timeout management, the following could be used instead:
  setTimeout(() => {
    reject(new ExtensionError("timed out"));
   }, 5000);
});
```

>     2. How would we actually go about converting the first two entries in a Uint8array that represent byte values into a 16-bit integer, which would represent the "length" field? The `dns-packet` module that we use (and slightly modify) [appears to call](https://github.com/mafintosh/dns-packet/blob/master/index.js#L1508-L1513) a Node.js function called [Buffer.readUInt16BE() ](https://www.geeksforgeeks.org/node-js-buffer-readuint16be-method/) to do this in their code for decoding TCP responses, but I'm not sure if we can do that here.

Since your whole project is browserified, you can use some Node.js APIs anywhere in your code, including the `Buffer` API (which I previously referred to at https://github.com/mozilla-extensions/dnssec-interference/pull/5#discussion_r467318024 ).

The `Buffer` API is Node.js-only, but in browsers the `DataView` API offers similar capabilities: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView

Even without a `Buffer` or `DataView` API, you could also get the first two bytes from a `Uint8Array` and then use (bitwise) arithmetic to calculate the value (bitshift = `<<`, XOR = `|`).

Reply:
I took a first pass at simplifying the `experiments.tcpsocket` API in 19ef3f9c678d6c62123b3f6eb59190c0e0576ea1, as per your helpful comments above. I roughly followed the pseudocode you wrote and removed most of the code borrowed from the test file. I have two questions about my code and how it can be simplified further:

1. The `socket.open()` function is asynchronous, so we can't assume that a TCP connection has been opened after the call is made. Furthermore, if you try to send a query before `socket.readyState = "open"`, then an error is thrown. Thus, we need some clean way of waiting until `socket.readyState = "open"` before we call `socket.send()`. In the code I have now, I await a Promise that is resolved when either the `socket.onopen` or `socket.onerror` event is fired. However, it doesn't seem very clean, and I end up setting event listeners again once we've sent the query so that no weird reference errors come up to old "resolve" and "reject" variables. In short, do you think there's a cleaner way to wait until the socket has either opened or an error has been fired before sending a query? The test file does this in a more complicated way [by storing a Promise in a variable](https://searchfox.org/mozilla-central/source/dom/network/tests/test_tcpsocket_client_and_server_basics.js#120-131).
2. If we get an `error` event before an `open` event, then does it make more sense to throw an error or call `reject()`? At the beginning of `sendDNSQuery()`, I use throw an error, but I wonder if it's more idiomatic to call `reject()` when we reach a bad state. In general, what is the purpose of `reject` versus throwing an error?

Reply:
> I took a first pass at simplifying the `experiments.tcpsocket` API in [19ef3f9](https://github.com/mozilla-extensions/dnssec-interference/commit/19ef3f9c678d6c62123b3f6eb59190c0e0576ea1), as per your helpful comments above. I roughly followed the pseudocode you wrote and removed most of the code borrowed from the test file. I'm able to send/receive queries over TCP again but I have two questions about my code and how it can be simplified further:
> 
>     1. By calling `new TCPSocket()`, a TCP connection is opened asynchronously. If you try to send a query immediately after `new TCPSocket()`, an error is usually thrown because`socket.readyState != "open"`. Thus, we need some clean way of waiting until we either get an `open` event or an `error` event before we think about calling `socket.send()`. In the code I have now, I await a Promise that is resolved when either the `socket.onopen` or `socket.onerror` event is fired. However, it doesn't seem very clean, and I end up setting event listeners again once we've sent the query so that no weird reference errors come up to old "resolve" and "reject" variables. In short, do you think there's a cleaner way to wait until the socket has either opened or an error has been fired before calling `socket.send()`? The test file does this in a more complicated way [by storing a Promise in a variable](https://searchfox.org/mozilla-central/source/dom/network/tests/test_tcpsocket_client_and_server_basics.js#120-131).

Like what I showed before, except with the following addition:

```
socket.onopen = () => {
  socket.onopen = null; // Not sure if needed, but doesn't hurt.
  socket.send(...);
};
```

I'm using the above snippet because there is only a need to send data once, and the onerror handler can then be reused.

>     2. If we get an `error` event before an `open` event, then does it make more sense to throw an error or call `reject()`? At the beginning of `sendDNSQuery()`, I throw an error, but I wonder if it's more idiomatic to call `reject()` when we reach a bad state. In general, what is the purpose of `reject` versus throwing an error?

As explained in a comment in a newer version of this patch, you cannot use `throw` from an event handler. The `Promise` constructor is used because you don't have a promise to `await` on, and that means that you explicitly need to call resolve or reject to settle the promise.

Reply:
Addressed in caf03b1 and 62d8337